keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,935 @@
+proper network planning before adding new ranges to the Service IP pool.
+
+## Design Details
+
+### Current implementation details 
+
+[Kubernetes
+Services](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#service-v1-core) need
+to be able to dynamically allocate resources from a predefined range/pool for ClusterIPs.
+
+ClusterIP is the IP address of the service and is usually assigned randomly. If an address is
+specified manually, is in-range, and is not in use, it will be allocated to the service; otherwise
+creation of the service will fail. The Service ClusterIP range is defined in the kube-apiserver with
+the following flags:
+
+```
+--service-cluster-ip-range string
+A CIDR notation IP range from which to assign service cluster IPs. This must not overlap with any IP ranges assigned to nodes or pods. Max of two dual-stack CIDRs is allowed.
+```
+
+And in the controller-manager:
+
+```
+--service-cluster-ip-range string
+CIDR Range for Services in cluster. Requires --allocate-node-cidrs to be true
+```
+
+The allocator is implemented using a bitmap that is serialized and stored in an ["opaque" API
+object](https://github.com/kubernetes/kubernetes/blob/b246220/pkg/apis/core/types.go#L5311).
+
+```go
+// RangeAllocation is an opaque API object (not exposed to end users) that can be persisted to record
+// the global allocation state of the cluster. The schema of Range and Data generic, in that Range
+// should be a string representation of the inputs to a range (for instance, for IP allocation it
+// might be a CIDR) and Data is an opaque blob understood by an allocator which is typically a
+// binary range.  Consumers should use annotations to record additional information (schema version,
+// data encoding hints). A range allocation should *ALWAYS* be recreatable at any time by observation
+// of the cluster, thus the object is less strongly typed than most.
+type RangeAllocation struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+	// A string representing a unique label for a range of resources, such as a CIDR "10.0.0.0/8" or
+	// port range "10000-30000". Range is not strongly schema'd here. The Range is expected to define
+	// a start and end unless there is an implicit end.
+	Range string
+	// A byte array representing the serialized state of a range allocation. Additional clarifiers on
+	// the type or format of data should be represented with annotations. For IP allocations, this is
+	// represented as a bit array starting at the base IP of the CIDR in Range, with each bit representing
+	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
+	Data []byte
+}
+```
+
+To avoid leaking resources, they also need to implement to repair loops that run in the
+`controlplane` instance that keep the bitmaps in sync with the current Services allocated IPs.
+
+The order of the IP families defined in the service-cluster-ip-range flag defines the primary IP family used by the allocator.
```

Comment:
We always need the flags to bootstrap the cluster, someone has to create the initial IPRange
The default Service IPFamily depend on the flags, the allocator uses it when you create a service without enough information, but this can be easily changed with a webhook or an admission controller, setting the IPFamily, IPFamilyPolicy fields .... 
We already don't have consistency on configurations on multiple apiservers, with this new approach the service ranges will be consistent, but the default IP family will still be defined by the flag orders ... is it worth it to fix this?

Reply:
It would be nice to say "eventually those flags can go away" even if it takes a couple years.

We can make that a followup KEP, I suppose - don't need to do it here.  Just be clear that this KEP does not remove the need for those flags